### PR TITLE
Improve handling configuration settings

### PR DESCRIPTION
**Pull request by randy3k**
1. Allows settings to be passed via `workspace/didChangeConfiguration`. 
2. Additionally, `options(languageserver.xxxx)` has always higher precedence to allow projectwise override.
3. Also added setting `rich_documentaion` for less powerful markdown renderer.

Discussion: do we want to set `rich_documentaion` to default FALSE and enable it manually in `vscode-r-lsp`?

### CODE REVIEW

File: R/settings.R [hunk 1]
```diff
@@ -0,0 +1,50 @@
+Settings <- R6::R6Class("Settings",
+    private = list(
```
Comment (renkun-ken) [inline R/settings.R:2]: `formatting_style` seems missing here?

Comment (randy3k) [reply R/settings.R:2]: `formatting_style` is a function and cannot be passed from the workspace configuration. Not sure how we should position it.

Comment (renkun-ken) [reply R/settings.R:2]: I see.

**Review by renkun-ken — APPROVED** [review]
LGTM

**Comment by renkun-ken** [comment]
BTW, do we want to support [`workspace/didChangeConfiguration`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeConfiguration)?

**Comment by renkun-ken** [comment]
> Discussion: do we want to set `rich_documentaion` to default FALSE and enable it manually in `vscode-r-lsp`?

I guess we could consider this if more users complain about it?

**Comment by randy3k** [comment]
> BTW, do we want to support [`workspace/didChangeConfiguration`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeConfiguration)?

We have supported it already. https://github.com/REditorSupport/languageserver/blob/cd4cbe53c239b56f2608b0d840d3d09bdca4f4d8/R/handlers-workspace.R#L13

**Comment by randy3k** [comment]
> > Discussion: do we want to set `rich_documentaion` to default FALSE and enable it manually in `vscode-r-lsp`?
> 
> 
> 
> I guess we could consider this if more users complain about it?

We could also check [MarkdownClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#markdownClientCapabilities). I assume that vscode will return [marked](https://marked.js.org).

**Comment by renkun-ken** [comment]
Here is the initialization config from latest vscode:

<details>
```json
{
  "processId": 35022,
  "clientInfo": {
    "name": "Visual Studio Code",
    "version": "1.57.1"
  },
  "locale": "en",
  "rootPath": "/home/ken/test",
  "rootUri": "file:///home/ken/test",
  "capabilities": {
    "workspace": {
      "applyEdit": true,
      "workspaceEdit": {
        "documentChanges": true,
        "resourceOperations": [
          "create",
          "rename",
          "delete"
        ],
        "failureHandling": "textOnlyTransactional",
        "normalizesLineEndings": true,
        "changeAnnotationSupport": {
          "groupsOnLabel": true
        }
      },
      "didChangeConfiguration": {
        "dynamicRegistration": true
      },
      "didChangeWatchedFiles": {
        "dynamicRegistration": true
      },
      "symbol": {
        "dynamicRegistration": true,
        "symbolKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26
          ]
        },
        "tagSupport": {
          "valueSet": [
            1
          ]
        }
      },
      "codeLens": {
        "refreshSupport": true
      },
      "executeCommand": {
        "dynamicRegistration": true
      },
      "configuration": true,
      "workspaceFolders": true,
      "semanticTokens": {
        "refreshSupport": true
      },
      "fileOperations": {
        "dynamicRegistration": true,
        "didCreate": true,
        "didRename": true,
        "didDelete": true,
        "willCreate": true,
        "willRename": true,
        "willDelete": true
      }
    },
    "textDocument": {
      "publishDiagnostics": {
        "relatedInformation": true,
        "versionSupport": false,
        "tagSupport": {
          "valueSet": [
            1,
            2
          ]
        },
        "codeDescriptionSupport": true,
        "dataSupport": true
      },
      "synchronization": {
        "dynamicRegistration": true,
        "willSave": true,
        "willSaveWaitUntil": true,
        "didSave": true
      },
      "completion": {
        "dynamicRegistration": true,
        "contextSupport": true,
        "completionItem": {
          "snippetSupport": true,
          "commitCharactersSupport": true,
          "documentationFormat": [
            "markdown",
            "plaintext"
          ],
          "deprecatedSupport": true,
          "preselectSupport": true,
          "tagSupport": {
            "valueSet": [
              1
            ]
          },
          "insertReplaceSupport": true,
          "resolveSupport": {
            "properties": [
              "documentation",
              "detail",
              "additionalTextEdits"
            ]
          },
          "insertTextModeSupport": {
            "valueSet": [
              1,
              2
            ]
          }
        },
        "completionItemKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25
          ]
        }
      },
      "hover": {
        "dynamicRegistration": true,
        "contentFormat": [
          "markdown",
          "plaintext"
        ]
      },
      "signatureHelp": {
        "dynamicRegistration": true,
        "signatureInformation": {
          "documentationFormat": [
            "markdown",
            "plaintext"
          ],
          "parameterInformation": {
            "labelOffsetSupport": true
          },
          "activeParameterSupport": true
        },
        "contextSupport": true
      },
      "definition": {
        "dynamicRegistration": true,
        "linkSupport": true
      },
      "references": {
        "dynamicRegistration": true
      },
      "documentHighlight": {
        "dynamicRegistration": true
      },
      "documentSymbol": {
        "dynamicRegistration": true,
        "symbolKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26
          ]
        },
        "hierarchicalDocumentSymbolSupport": true,
        "tagSupport": {
          "valueSet": [
            1
          ]
        },
        "labelSupport": true
      },
      "codeAction": {
        "dynamicRegistration": true,
        "isPreferredSupport": true,
        "disabledSupport": true,
        "dataSupport": true,
        "resolveSupport": {
          "properties": [
            "edit"
          ]
        },
        "codeActionLiteralSupport": {
          "codeActionKind": {
            "valueSet": [
              "",
              "quickfix",
              "refactor",
              "refactor.extract",
              "refactor.inline",
              "refactor.rewrite",
              "source",
              "source.organizeImports"
            ]
          }
        },
        "honorsChangeAnnotations": false
      },
      "codeLens": {
        "dynamicRegistration": true
      },
      "formatting": {
        "dynamicRegistration": true
      },
      "rangeFormatting": {
        "dynamicRegistration": true
      },
      "onTypeFormatting": {
        "dynamicRegistration": true
      },
      "rename": {
        "dynamicRegistration": true,
        "prepareSupport": true,
        "prepareSupportDefaultBehavior": 1,
        "honorsChangeAnnotations": true
      },
      "documentLink": {
        "dynamicRegistration": true,
        "tooltipSupport": true
      },
      "typeDefinition": {
        "dynamicRegistration": true,
        "linkSupport": true
      },
      "implementation": {
        "dynamicRegistration": true,
        "linkSupport": true
      },
      "colorProvider": {
        "dynamicRegistration": true
      },
      "foldingRange": {
        "dynamicRegistration": true,
        "rangeLimit": 5000,
        "lineFoldingOnly": true
      },
      "declaration": {
        "dynamicRegistration": true,
        "linkSupport": true
      },
      "selectionRange": {
        "dynamicRegistration": true
      },
      "callHierarchy": {
        "dynamicRegistration": true
      },
      "semanticTokens": {
        "dynamicRegistration": true,
        "tokenTypes": [
          "namespace",
          "type",
          "class",
          "enum",
          "interface",
          "struct",
          "typeParameter",
          "parameter",
          "variable",
          "property",
          "enumMember",
          "event",
          "function",
          "method",
          "macro",
          "keyword",
          "modifier",
          "comment",
          "string",
          "number",
          "regexp",
          "operator"
        ],
        "tokenModifiers": [
          "declaration",
          "definition",
          "readonly",
          "static",
          "deprecated",
          "abstract",
          "async",
          "modification",
          "documentation",
          "defaultLibrary"
        ],
        "formats": [
          "relative"
        ],
        "requests": {
          "range": true,
          "full": {
            "delta": true
          }
        },
        "multilineTokenSupport": false,
        "overlappingTokenSupport": false
      },
      "linkedEditingRange": {
        "dynamicRegistration": true
      }
    },
    "window": {
      "showMessage": {
        "messageActionItem": {
          "additionalPropertiesSupport": true
        }
      },
      "showDocument": {
        "support": true
      },
      "workDoneProgress": true
    },
    "general": {
      "regularExpressions": {
        "engine": "ECMAScript",
        "version": "ES2020"
      },
      "markdown": {
        "parser": "marked",
        "version": "1.1.0"
      }
    }
  },
  "trace": "off",
  "workspaceFolders": [
    {
      "uri": "file:///home/ken/test",
      "name": "test"
    }
  ]
}
```
</details>

**Comment by renkun-ken** [comment]
> > > Discussion: do we want to set `rich_documentaion` to default FALSE and enable it manually in `vscode-r-lsp`?
> > 
> > 
> > I guess we could consider this if more users complain about it?
> 
> We could also check [MarkdownClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#markdownClientCapabilities). I assume that vscode will return [marked](https://marked.js.org).

Yes, it would be much nicer. What do other clients show?

**Comment by randy3k** [comment]
I believe that most other clients do not return anything.

File: R/completion.R [hunk 2]
```diff
@@ -458,7 +458,7 @@ completion_reply <- function(id, uri, workspace, document, point, capabilities)
 
     t0 <- Sys.time()
     snippet_support <- isTRUE(capabilities$completionItem$snippetSupport) &&
-        getOption("languageserver.snippet_support", TRUE)
+        lsp_settings$get("snippet_support")
```
Comment (renkun-ken) [inline R/completion.R:461]: Is `lsp_settings` a big redundant? Is `settings` easily confused with something similar?

Comment (randy3k) [reply R/completion.R:461]: It was to differentiate from https://github.com/REditorSupport/languageserver/blob/master/R/handlers-workspace.R#L14

Comment (renkun-ken) [reply R/completion.R:461]: I see.

**Comment by randy3k** [comment]
@renkun-ken it seems that there is some problem to enable `lint_cache` of `lintr`. Specially, running `lintr` via `callr` with `cache = TRUE` seems to trigger this error:

```
[2021-06-22 07:36:06.604] diagnostics_task: <callr_status_error: callr subprocess failed: cannot open the connection>
 in process 
-->
<callr_remote_error in gzfile(file, "wb"):
 cannot open the connection>
 in process 5116 
```

**Comment by renkun-ken** [comment]
Thanks for the fix. Not sure about the cache problem either.

File: README.md [hunk 3]
```diff
@@ -109,22 +109,27 @@ These editors are supported by installing the corresponding package.
 
 ## Settings
 
-`languageserver` exposes the following settings via `workspace/didChangeConfiguration`
+`languageserver` exposes the following settings via LSP configuration.
+
+settings | default | description
+----     | -----   | -----
+`debug`  | `false` | increase verbosity for debug purpose
+`log_file` | `nil` | file to log debug messages, fallback to stderr if empty
```
Comment (eitsupi) [inline README.md:117]: it is nil, not null?

Comment (eitsupi) [reply README.md:117]: Thanks!

Comment (randy3k) [reply README.md:117]: You are right. Too much golang this week.

**Pull request merged**